ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/AutoCreateIcebergTagAction.java

```diff
@@ -0,0 +1,197 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.server.optimizing.maintainer;
+
+import com.netease.arctic.table.TableProperties;
+import com.netease.arctic.table.TagTriggerPeriod;
+import com.netease.arctic.utils.CompatiblePropertyUtil;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Map;
+
+public class AutoCreateIcebergTagAction {
+  private static final Logger LOG = LoggerFactory.getLogger(AutoCreateIcebergTagAction.class);
+
+  private final Table table;
+  private final TagConfig tagConfig;
+  private final LocalDateTime now;
+
+  public AutoCreateIcebergTagAction(Table table, LocalDateTime now) {
+    this.table = table;
+    this.tagConfig = TagConfig.fromTableProperties(table.properties());
+    this.now = now;
+  }
+
+  public void execute() {
+    if (!tagConfig.isAutoCreateTag()) {
+      return;
+    }
+    if (tagExist()) {
+      LOG.debug("{} find expect tag, skip", table.name());
+      return;
+    }
+    boolean success = createTag();
+    LOG.info("{} tag creation {}", table.name(), success ? "succeed" : "skipped");
+  }
+
+  private boolean tagExist() {
+    if (TagTriggerPeriod.DAILY.propertyName().equals(tagConfig.getTriggerPeriod())) {
+      return findTagOfToday() != null;
+    } else {
+      throw new IllegalArgumentException(
+          "unsupported trigger period " + tagConfig.getTriggerPeriod());
+    }
+  }
+
+  private String findTagOfToday() {
+    String name = generateTagName();
+    return table.refs().entrySet().stream()
+        .filter(entry -> entry.getValue().isTag())
+        .map(Map.Entry::getKey)
+        .filter(name::equals)
+        .findFirst()
+        .orElse(null);
+  }
+
+  private boolean createTag() {
+    Snapshot snapshot = findSnapshot(table, getTagTriggerTime());
+    if (snapshot == null) {
+      LOG.info("{} no snapshot found at {}", this.table.name(), getTagTriggerTime());
+      return false;
+    }
+    String newTagName = generateTagName();
+    table.manageSnapshots().createTag(newTagName, snapshot.snapshotId()).commit();
+    LOG.info(
+        "{} create tag {} on snapshot {} {}",
+        this.table.name(),
+        newTagName,
+        snapshot.snapshotId(),
+        snapshot.timestampMillis());
+    return true;
+  }
+
+  private String generateTagName() {
+    if (TagTriggerPeriod.DAILY.propertyName().equals(tagConfig.getTriggerPeriod())) {
+      String tagFormat = tagConfig.getTagFormat();
+      return now.minusDays(1).format(DateTimeFormatter.ofPattern(tagFormat));
+    } else {
+      throw new IllegalArgumentException(
+          "unsupported trigger period " + tagConfig.getTriggerPeriod());
+    }
+  }
+
+  private long getTagTriggerTime() {
```

Comment:
It feels more reasonable to move this logic to `TagTriggerPeriod`

Reply:
I will have a try.